src/common/storage_map.rs

```diff
@@ -1,57 +1,71 @@
-use fuel_storage::Storage;
+use fuel_storage::{Mappable, StorageInspect, StorageMutate};
 
 use alloc::borrow::Cow;
 use hashbrown::HashMap;
 
 #[derive(Debug)]
-pub struct StorageMap<Key, Value> {
-    map: HashMap<Key, Value>,
+pub struct StorageMap<Type: Mappable> {
+    map: HashMap<Type::Key, Type::GetValue>,
 }
 
-impl<Key, Value> Default for StorageMap<Key, Value> {
+impl<Type: Mappable> Default for StorageMap<Type> {
     fn default() -> Self {
         Self::new()
     }
 }
 
-impl<Key, Value> StorageMap<Key, Value> {
+impl<Type: Mappable> StorageMap<Type> {
     pub fn new() -> Self {
         Self {
-            map: HashMap::<Key, Value>::new(),
+            map: Default::default(),
         }
     }
 }
 
-impl<Key, Value> Storage<Key, Value> for StorageMap<Key, Value>
+impl<Type> StorageInspect<Type> for StorageMap<Type>
 where
-    Key: Eq + core::hash::Hash + Clone,
-    Value: Clone,
+    Type: Mappable,
+    Type::Key: Eq + core::hash::Hash + Clone,
+    Type::GetValue: Clone,
 {
     type Error = core::convert::Infallible;
 
-    fn insert(&mut self, key: &Key, value: &Value) -> Result<Option<Value>, Self::Error> {
-        self.map.insert(key.clone(), value.clone());
-        let v = Some(value.clone());
-        Ok(v)
-    }
-
-    fn remove(&mut self, key: &Key) -> Result<Option<Value>, Self::Error> {
-        let value = self.map.remove(key);
-        Ok(value)
-    }
-
-    fn get(&self, key: &Key) -> Result<Option<Cow<Value>>, Self::Error> {
+    fn get(&self, key: &Type::Key) -> Result<Option<Cow<Type::GetValue>>, Self::Error> {
         let result = self.map.get(key);
-        let value = result.map(|value| Cow::Borrowed(value));
+        let value = result.map(Cow::Borrowed);
         Ok(value)
     }
 
-    fn contains_key(&self, key: &Key) -> Result<bool, Self::Error> {
+    fn contains_key(&self, key: &Type::Key) -> Result<bool, Self::Error> {
         let contains = self.map.contains_key(key);
         Ok(contains)
     }
 }
 
+impl<Type> StorageMutate<Type> for StorageMap<Type>
+where
+    Type: Mappable,
+    Type::Key: Eq + core::hash::Hash + Clone,
+    Type::SetValue: Clone,
+    Type::GetValue: Clone + From<Type::SetValue>,
+{
+    fn insert(
+        &mut self,
+        key: &Type::Key,
+        value: &Type::SetValue,
+    ) -> Result<Option<Type::GetValue>, Self::Error> {
+        let previous = self.map.remove(key);
+
+        self.map.insert(key.clone(), value.clone().into());
+        Ok(previous)
```

Comment:
It looks like you are changing the behaviour of the insert to return the previous value at `key`. At the time of writing the original `insert`, there was no standard definition, AFAIK. Is returning the previous value the proper idiom?

Reply:
Yes, it is how [`HashMap` works](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.insert) in Rust. Insert value `X` and return value `Some(X)` looks strange=D